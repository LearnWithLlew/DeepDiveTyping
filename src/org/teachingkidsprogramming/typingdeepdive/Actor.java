package org.teachingkidsprogramming.typingdeepdive;

import java.awt.Dimension;
import java.awt.Graphics;

public interface Actor
{
  void advanceClock();
  boolean processLetter(char letter);
  void paint(Graphics g, Dimension size);
}
