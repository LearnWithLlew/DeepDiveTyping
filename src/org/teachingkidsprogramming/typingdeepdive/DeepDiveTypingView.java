package org.teachingkidsprogramming.typingdeepdive;

import java.awt.Color;
import java.awt.Dimension;
import java.awt.Graphics;

import javax.swing.JPanel;

public class DeepDiveTypingView extends JPanel
{
  private DeepDiveTypingGame deepDive;
  public DeepDiveTypingView(DeepDiveTypingGame deepDive)
  {
    this.deepDive = deepDive;
    this.setPreferredSize(new Dimension(640, 480));
    this.setBackground(Color.BLUE);
    this.addKeyListener(deepDive);
    this.setFocusable(true);
  }
  @Override
  public void paint(Graphics g)
  {
    super.paint(g);
    for (Actor shark : deepDive.actors)
    {
      shark.paint(g, getSize());
    }
  }
}
