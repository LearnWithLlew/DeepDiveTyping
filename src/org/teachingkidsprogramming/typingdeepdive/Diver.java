package org.teachingkidsprogramming.typingdeepdive;

import java.awt.Dimension;
import java.awt.Graphics;
import java.awt.Image;

import javax.swing.ImageIcon;

public class Diver implements Actor
{
  private static Image image;
  private int          frames;
  @Override
  public void advanceClock()
  {
    frames++;
  }
  @Override
  public boolean processLetter(char letter)
  {
    return false;
  }
  @Override
  public void paint(Graphics g, Dimension size)
  {
    int imageCount = 8;
    Image image = loadImage();
    int width = image.getWidth(null) / imageCount;
    int height = image.getHeight(null);
    int start = ((frames / 4) % imageCount) * width;
    int y = (size.height - height) / 2;
    g.drawImage(image, 0, y, width, y + height, start, 0, start + width, height, null);
  }
  public static Image loadImage()
  {
    if (image == null)
    {
      image = new ImageIcon(Diver.class.getResource("diver.png")).getImage();
    }
    return image;
  }
}
