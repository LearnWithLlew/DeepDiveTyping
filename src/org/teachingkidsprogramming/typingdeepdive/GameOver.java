package org.teachingkidsprogramming.typingdeepdive;

import java.awt.Dimension;
import java.awt.Graphics;

public class GameOver implements Actor
{
  private DeepDiveTypingGame game;
  public GameOver(DeepDiveTypingGame game)
  {
    this.game = game;
    game.selected = this;
    game.timer.stop();
  }
  @Override
  public void advanceClock()
  {
  }
  @Override
  public boolean processLetter(char letter)
  {
    if (letter == 's')
    {
      game.start();
      game.timer.start();
      return true;
    }
    return false;
  }
  @Override
  public void paint(Graphics g, Dimension size)
  {
    StartGame.paintText(g, size, "Game Over \n re(S)tart Game");
  }
}
