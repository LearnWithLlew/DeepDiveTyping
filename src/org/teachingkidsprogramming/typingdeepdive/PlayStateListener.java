package org.teachingkidsprogramming.typingdeepdive;

public interface PlayStateListener
{
  void playStateChanged(Shark shark);
}
