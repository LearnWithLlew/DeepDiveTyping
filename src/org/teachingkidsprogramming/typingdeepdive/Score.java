package org.teachingkidsprogramming.typingdeepdive;

public class Score
{
  private int sharks;
  private int batch;
  public void scoreKill(Shark shark)
  {
    setSharks(getSharks() + 1);
  }
  public int getSharks()
  {
    return sharks;
  }
  public void setSharks(int sharks)
  {
    this.sharks = sharks;
  }
  public void scoreNewBatch()
  {
    setBatch(getBatch() + 1);
  }
  public int getBatch()
  {
    return batch;
  }
  public void setBatch(int batch)
  {
    this.batch = batch;
  }
}
