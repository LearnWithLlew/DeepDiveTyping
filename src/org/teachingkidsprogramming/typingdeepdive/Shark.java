package org.teachingkidsprogramming.typingdeepdive;

import java.awt.Color;
import java.awt.Dimension;
import java.awt.Graphics;
import java.awt.Point;
import java.util.ArrayList;

import com.spun.util.ArrayUtils;

public class Shark
{
  public enum PlayState {
    Active, Selected, Killed, Killing
  };
  public static final int              speed     = 1000 / (15 * 25);
  private String                       word;
  private int                          typed     = 0;
  private Dimension                    outer     = new Dimension(200, 50);
  private int                          diver     = 75;
  private Point                        topRight;
  private PlayState                    state;
  private ArrayList<PlayStateListener> listeners = new ArrayList<PlayStateListener>();
  public Shark()
  {
    this("Samantha", 200);
  }
  public Shark(String word, int y)
  {
    this.word = word;
    this.setState(PlayState.Active);
    this.topRight = new Point(1000, y);
  }
  public void paint(Graphics g, Dimension box)
  {
    int startX = box.width - outer.width;
    int length = startX - diver;
    double xScale = length / 1000.0;
    double yScale = box.getHeight() / 1000.0;
    int xPosition = (int) ((xScale * topRight.x) + diver);
    int yPosition = (int) (yScale * topRight.y);
    g.setColor(Color.WHITE);
    g.drawString(word, xPosition, yPosition);
    g.setColor(Color.RED);
    g.drawString(getCompleted(), xPosition, yPosition);
  }
  public String getCompleted()
  {
    return word.substring(0, typed);
  }
  public void advanceClock()
  {
    if (isFinished()) { return; }
    topRight.x -= speed;
    if (topRight.x <= 0)
    {
      setState(PlayState.Killing);
    }
  }
  public void processLetter(char letter)
  {
    if (isFinished()) { return; }
    char currentLetter = word.charAt(typed);
    if (currentLetter == letter)
    {
      setState(PlayState.Selected);
      typed++;
      if (typed == word.length())
      {
        setState(PlayState.Killed);
      }
    }
  }
  public boolean isFinished()
  {
    return ArrayUtils.contains(new PlayState[]{PlayState.Killed, PlayState.Killing}, getState());
  }
  @Override
  public String toString()
  {
    String template = "Shark{word:%s,completed:%s,topRight:%s,state:%s}";
    return String.format(template.replace(',', '\n'), word, getCompleted(), topRight, getState());
  }
  public PlayState getState()
  {
    return state;
  }
  public void addListener(PlayStateListener playStateListener)
  {
    listeners.add(playStateListener);
  }
  private void setState(PlayState state)
  {
    if (this.state != state)
    {
      this.state = state;
      for (PlayStateListener l : listeners)
      {
        l.playStateChanged(this);
      }
    }
  }
}
