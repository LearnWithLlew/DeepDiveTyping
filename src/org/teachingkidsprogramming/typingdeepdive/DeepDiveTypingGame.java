package org.teachingkidsprogramming.typingdeepdive;

import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.KeyEvent;
import java.awt.event.KeyListener;
import java.util.ArrayList;

import javax.swing.JPanel;
import javax.swing.Timer;

import org.teachingkidsprogramming.typingdeepdive.Shark.PlayState;

import com.spun.util.WindowUtils;

public class DeepDiveTypingGame implements KeyListener, PlayStateListener
{
  public JPanel           view   = new DeepDiveTypingView(this);
  public ArrayList<Actor> actors = new ArrayList<Actor>();
  public Score            score;
  public Timer            timer;
  public Actor            selected;
  private void launchWindow()
  {
    showStartPanel();
    WindowUtils.testPanel(view);
  }
  private void showStartPanel()
  {
    actors.add(new StartGame(this));
    startClock();
  }
  private void startClock()
  {
    int delay = 1000 / 25; //milliseconds
    ActionListener taskPerformer = new ActionListener()
    {
      @Override
      public void actionPerformed(ActionEvent e)
      {
        advanceClock();
        view.repaint();
      }
    };
    timer = new Timer(delay, taskPerformer);
    timer.start();
  }
  protected void advanceClock()
  {
    for (Actor shark : actors.toArray(new Actor[0]))
    {
      shark.advanceClock();
    }
  }
  public static void main(String[] args)
  {
    new DeepDiveTypingGame().launchWindow();
  }
  private void processLetter(char letter)
  {
    if (selected != null)
    {
      selected.processLetter(letter);
      return;
    }
    for (Actor shark : actors.toArray(new Actor[0]))
    {
      if (selected == null)
      {
        if (shark.processLetter(letter)) { return; }
      }
    }
  }
  @Override
  public void keyTyped(KeyEvent e)
  {
    processLetter(e.getKeyChar());
  }
  @Override
  public void keyPressed(KeyEvent e)
  {
    //do nothing
  }
  @Override
  public void keyReleased(KeyEvent e)
  {
    //do nothing
  }
  @Override
  public void playStateChanged(Shark shark)
  {
    PlayState state = shark.getState();
    switch (state)
    {
      case Killed :
        removeShark(shark);
        break;
      case Killing :
        actors.add(new GameOver(this));
        break;
      case Selected :
        selected = shark;
        break;
      default :
        break;
    }
  }
  public void removeShark(Shark shark)
  {
    score.scoreKill(shark);
    remove(shark);
  }
  public void remove(Actor shark)
  {
    actors.remove(shark);
    selected = null;
    if (isEmpty())
    {
      actors.add(new SharkBatch(this));
    }
    else
    {
      System.out.println(actors.toString());
    }
  }
  private boolean isEmpty()
  {
    return actors.isEmpty() || (actors.size() == 1 && actors.get(0) instanceof Diver);
  }
  public void addDiver()
  {
    boolean contains = false;
    for (Actor a : actors)
    {
      contains |= a instanceof Diver;
    }
    if (!contains)
    {
      actors.add(new Diver());
    }
  }
  public void start()
  {
    score = new Score();
    actors.clear();
    selected = null;
    actors.add(new SharkBatch(this));
    actors.add(new Diver());
  }
}
