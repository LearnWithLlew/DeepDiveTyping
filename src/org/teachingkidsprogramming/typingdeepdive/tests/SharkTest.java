package org.teachingkidsprogramming.typingdeepdive.tests;

import org.approvaltests.Approvals;
import org.junit.Test;
import org.teachingkidsprogramming.typingdeepdive.Shark;

public class SharkTest
{
  @Test
  public void testSharkEatsYou() throws Exception
  {
    Shark shark = new Shark("loser", 200);
    advanceTillEnd(shark);
    Approvals.verify("" + shark);
  }
  private void advanceTillEnd(Shark shark)
  {
    for (int i = 0; i < 1000; i++)
    {
      shark.advanceClock();
    }
  }
  @Test
  public void testFirstLetter() throws Exception
  {
    Shark shark = new Shark("word", 200);
    shark.processLetter('w');
    Approvals.verify("" + shark);
  }
  @Test
  public void testWordCompleted() throws Exception
  {
    Shark shark = new Shark("word", 200);
    PlayStateRecorder listener = new PlayStateRecorder();
    shark.addListener(listener);
    processLetters("word", shark);
    Approvals.verify(shark + "\n" + listener.log);
  }
  private void processLetters(String string, Shark shark)
  {
    for (char c : string.toCharArray())
    {
      shark.processLetter(c);
    }
  }
}
