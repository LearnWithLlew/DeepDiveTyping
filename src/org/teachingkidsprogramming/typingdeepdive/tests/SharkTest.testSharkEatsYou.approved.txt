Shark{word:loser
completed:
topRight:java.awt.Point[x=0,y=200]
state:Killing}